main/src/library/Set.flix

```diff
@@ -264,46 +256,43 @@ namespace Set {
     /// The function `f` must be pure.
     ///
     pub def filter(f: a -> Bool, xs: Set[a]): Set[a] =
-        let Set(s) = xs;
-            Set(List.filter(f, s))
+        foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs)
 
     ///
     /// Returns the result of applying `f` to every element in `xs`.
     ///
     /// Note: The returned set may be smaller than `xs`.
     ///
-    pub def map(f: a -> b & e, xs: Set[a]): Set[b] & e = match xs {
-        case Set(Nil) => Set(Nil)
-        case Set(x :: rs) => insert(f(x), map(f, Set(rs)))
-    }
+    pub def map(f: a -> b & e, xs: Set[a]): Set[b] & e =
+        foldLeft((acc, x) -> insert(f(x), acc), empty(), xs)
```

Comment:
This could probably be implemented more efficient on the tree itself? (You would skip rotations! Seems like an interesting experiment!!!)

Reply:
Makes sense, as we are only changing the values here. Good call!
I will add it to my list of experiment.

Reply:
I guess there is an optimization hidden somewhere in this, but I suspect that it would only hold if f is an order isomorphism(?)